### PR TITLE
[MM-31094] Replication Lag

### PR DESCRIPTION
**Pull request by mkraft**
…ion lag. Adds some lazy lookups to fallback to master if results are not found.

#### Summary

Reads from master in some cases. Adds test that simulate replication lag. Adds dev tooling to develop with replication lag.

Related PR: https://github.com/mattermost/enterprise/pull/919

#### Ticket Link

https://mattermost.atlassian.net/browse/MM-31094

#### Release Note

```release-note
Fixes bugs related to replication lag for Enterprise Edition instances configured to use read replicas.
```

### CODE REVIEW

File: testlib/helper.go
Comment (mkraft) [inline testlib/helper.go:29]: Need to remove this.

**Review by hahmadia — COMMENTED** [review]
Small comment + Questions asked over DM’s. Thanks

**Review by sbishel — APPROVED** [review]
I like that it is done consistently now. Looks good.

File: store/sqlstore/store.go
Comment (agnivade) [inline store/sqlstore/store.go:1375]: It looks like this is only to be used for testing. Wondering if we can move it to a `_test.go` file instead? Otherwise it will be part of the built binary too.

File: app/session.go
Comment (agnivade) [inline app/session.go:88]: Nice, you have already solved https://mattermost.atlassian.net/browse/MM-32620 :)

File: store/sqlstore/group_store.go
Comment (agnivade) [inline store/sqlstore/group_store.go:737]: Were these queries only used by the LDAP sync code, and therefore always pointing to master?

File: store/sqlstore/session_store.go
Comment (agnivade) [inline store/sqlstore/session_store.go:62]: Can you clarify which of the 3 issues is this fix for?

**Review by agnivade — COMMENTED** [review]
Great work there Martin.

Can you expand a bit on the changes that were made for the "login of normal user" fix? From what I understand of the issue- 
- The login request hits one server and creates the session.
- The second request hits a different server which does not have the item in the session cache.
- It misses the cache, and queries replica, and therefore misses the item and throws an error.

That's exactly what https://mattermost.atlassian.net/browse/MM-31252 and https://mattermost.atlassian.net/browse/MM-32620 is about, and the fix seems to be to query master for session cache misses, which you have already done. I am wondering if there were any other items that are needed to fix this?

Lastly, I made a list of all the app methods that now query master. It would be great if you could kindly categorize the changes into which of the 3 fixes is it part of. This is so that I completely understand which change fixes which bug. My apologies for not being able to understand this by just reading the code.

JoinDefaultChannels - User().Get()
AddUserToChannel - Team().GetMember()
CreatePost- Post().Get() - This one is for the reply to post.
GetSession - Query master on cache miss - This one is for user login I believe.
UpdateTeamMemberRoles - Team().GetMember()
AddUserToTeam - User().Get()
JoinUserToTeam- Team().GetMember()
GetTeamMember- Team().GetMember()
GetUser - User().Get() - I am especially curious about this one.

**Review by hahmadia — APPROVED** [review]
Had a meeting with Martin and explained everything to me. Thanks!

File: store/sqlstore/group_store.go
Comment (mkraft) [reply store/sqlstore/group_store.go:737]: Correct.

File: utils/testutils/testutils.go
Comment (agnivade) [inline utils/testutils/testutils.go:79]: That's not quite what I meant, but now that we have done this, it's not a big deal. :P

Comment (mkraft) [reply utils/testutils/testutils.go:79]: What did you have in mind Agniva?

Comment (agnivade) [reply utils/testutils/testutils.go]: Ah I see :)

Well, I wanted the method to be moved to a `_test.go` file under the same `sqlstore` package. So there would be no code changes from the callsite, but the method wouldn't appear in built binaries, and is semantically more clear.

**Comment by mkraft** [comment]
Unfortunately putting this back in WIP because I'm getting flaky tests running `MM_SQLSETTINGS_DRIVERNAME=mysql make test-replication-lag`.

**Comment by agnivade** [comment]
Hi @mkraft - @lindy65 is facing some issues on her Rainforest servers due to https://mattermost.atlassian.net/browse/MM-32620, which is already covered in this PR. That's why I didn't want to send a PR myself and undo the work that you have done.

Wondering if it would be okay if I send a PR separately to query master for session cache misses and help unblock Lindy? Otherwise, if you are nearly done with this, I wouldn't want to mess up your PR.

**Comment by mkraft** [comment]
> Hi @mkraft - @lindy65 is facing some issues on her Rainforest servers due to https://mattermost.atlassian.net/browse/MM-32620, which is already covered in this PR. That's why I didn't want to send a PR myself and undo the work that you have done.
> 
> Wondering if it would be okay if I send a PR separately to query master for session cache misses and help unblock Lindy? Otherwise, if you are nearly done with this, I wouldn't want to mess up your PR.

@agnivade Yes please feel free to send a separate PR. I'll just merge it in and remove that part of this PR if it's in first. Same goes for any part that will be following the "proper" fixes for these issues (rather than the temporary fixes found in this PR).

File: store/sqlstore/session_store.go
Comment (mkraft) [inline store/sqlstore/session_store.go:86]: TODO: Revert this change.

**Comment by mkraft** [comment]
> Great work there Martin.
> 
> Can you expand a bit on the changes that were made for the "login of normal user" fix? From what I understand of the issue-
> 
> * The login request hits one server and creates the session.
> * The second request hits a different server which does not have the item in the session cache.
> * It misses the cache, and queries replica, and therefore misses the item and throws an error.
> 
> That's exactly what https://mattermost.atlassian.net/browse/MM-31252 and https://mattermost.atlassian.net/browse/MM-32620 is about, and the fix seems to be to query master for session cache misses, which you have already done. I am wondering if there were any other items that are needed to fix this?
> 
> Lastly, I made a list of all the app methods that now query master. It would be great if you could kindly categorize the changes into which of the 3 fixes is it part of. This is so that I completely understand which change fixes which bug. My apologies for not being able to understand this by just reading the code.
> 
> **JoinDefaultChannels** - User().Get()
> **AddUserToChannel** - Team().GetMember()
> **CreatePost**- Post().Get() - This one is for the reply to post.
> **GetSession** - Query master on cache miss - This one is for user login I believe.
> **UpdateTeamMemberRoles** - Team().GetMember()
> **AddUserToTeam** - User().Get()
> **JoinUserToTeam**- Team().GetMember()
> **GetTeamMember**- Team().GetMember()
> **GetUser** - User().Get() - I am especially curious about this one.

Here's a summary of the changes as of now:

* (*App).JoinDefaultChannels gets the user from master
* (*App).AddUserToChannel gets the team member from master
* (*App).UpsertGroupSyncable never gets the channel from cache
* (*App).CreatePost gets the post from master
* (*App).CreateSession gets the session from master
* (*App).GetTeamMember gets the team member from master
* (*SqlGroupStore).TeamMembersToAdd always queries from master
* (*SqlGroupStore).ChannelMembersToAdd always queries from master

**Comment by agnivade** [comment]
Thanks Martin. My question was more on which change fixes which issue. If you could categorize them, I'll have a better idea of the scope of the changes.

I just need a mapping of your above line items to the following points below.

1. Very first login of an LDAP or SAML user
2. Login (not first) of any user
3. LDAP sync
4. Reply to post

**Comment by mkraft** [comment]
> Thanks Martin. My question was more on which change fixes which issue. If you could categorize them, I'll have a better idea of the scope of the changes.
> 
> I just need a mapping of your above line items to the following points below.
> 
> 1. Very first login of an LDAP or SAML user
> 2. Login (not first) of any user
> 3. LDAP sync
> 4. Reply to post

Gotcha. Here's the breakdown:

1. Very first login of an LDAP or SAML user
    * `(*App).AddUserToChannel`
    * `(*App).GetTeamMember`
    * `(*SqlGroupStore).TeamMembersToAdd`
    * `(*SqlGroupStore).ChannelMembersToAdd`
2. Login (not first) of any user
    * `(*App).CreateSession`
3. LDAP sync
    * `(*App).AddUserToChannel`
    * `(*App).GetTeamMember`
    * `(*SqlGroupStore).TeamMembersToAdd`
    * `(*SqlGroupStore).ChannelMembersToAdd`
4. Reply to post
    * `(*App).CreatePost`

**Comment by agnivade** [comment]
That's exactly what I needed. Thanks Martin!

And just before we merge this PR, please ping either me or @streamer45 or get a load-test done, just so we confirm that there is no regression perf wise. The LDAP paths won't be tested anyways, but for the other 2 cases, it'll be good to be thorough.

**Comment by mkraft** [comment]
@streamer45 or @agnivade Ready for perf. eval.

**Comment by agnivade** [comment]
Load tested with a 4M MySQL dataset with (1+1) DB setup. Did not observe any noticeable difference.

```
Comparison result:
Report: report_0_mysql_unbounded.md
Grafana Dashboard: http://34.227.75.47:3000/d/wTTxHa8Mz/comparison-0-mysql-unbounded
master:
  Type: unbounded
  DB Engine: mysql
  Supported Users: 4827
  Errors: 474
replica:
  Type: unbounded
  DB Engine: mysql
  Supported Users: 4824
  Errors: 2051
==================================================
```

replica-comp1

replica-comp2

/cc @streamer45

**Review by agnivade — APPROVED** [review]

**Review by furqanmlk — APPROVED** [review]
1. Reply to new post before replication completes. See no error. ✅
2. Login and clear cache before replication completes. Reload page. ✅
3. First login of LDAP user gets all expected memberships (groups, teams, and channels). ✅
4. LDAP sync adds the following memberships for q users where q > 1 (group, team, channel).
5. Clear all caches as system admin, reload the page and don't get logged out immediately. ✅
6. General smoke testing of the app does not trigger too many unnecessary extra reads against master (as monitored by the logging). ✅

**Comment by agnivade** [comment]
@lindy65 - Please feel free to check and let us know that the `/test post` issue is resolved now. No rush of course. Thanks1

**Comment by lindy65** [comment]
Thanks @agnivade - I will only be able to check the Rainforest servers once we have a v5.34 build.  I've tested on `master.test` server though and the command is failing intermittently (testing commands are enabled via the sys cons)

1. 
image

2. 
image

3. 
image

The command I'm using is: `/test post u=@user-1 p=user-1 c=~town-square t=ad-1 "message"`

**Comment by agnivade** [comment]
Can you show me the server logs of the error from the system console?

**Comment by lindy65** [comment]
Is this what you're looking for (just a snippet of where I could see "/test post")

```
{"level":"debug","ts":1615890046.831672,"caller":"app/command.go:437","msg":"Executing command","command":"test","user_id":"crh8mhddhbdjmk1fpk8e5r71tr"}
{"level":"error","ts":1615890046.8475707,"caller":"mlog/log.go:232","msg":"Command with a trigger of 'test' failed.","path":"/api/v4/commands/execute","request_id":"xt114giqwfb6idie8c9gc9t3qo","ip_addr":"105.226.243.4","user_id":"crh8mhddhbdjmk1fpk8e5r71tr","method":"POST","err_where":"command","http_code":500,"err_details":"Post \"http:\": http: no Host in request URL"}
{"level":"debug","ts":1615890046.847642,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"POST","url":"/api/v4/commands/execute","request_id":"xt114giqwfb6idie8c9gc9t3qo","status_code":"500"}
{"level":"debug","ts":1615890051.9275894,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"GET","url":"/","request_id":"4pcaij6z73ynjy5ch6md9hzbah","status_code":"200"}
{"level":"debug","ts":1615890053.3940399,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"POST","url":"/api/v4/channels/members/me/view","request_id":"8w75r1ijeid17fhu4z8du7zp7w","status_code":"200"}
{"level":"debug","ts":1615890053.463839,"caller":"app/analytics.go:29","msg":"More than limit users are on the system, intensive queries skipped","limit":2500}
{"level":"debug","ts":1615890053.4744482,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"GET","url":"/api/v4/plugins","request_id":"en6certx43ffd8mpqdizs34f4y","status_code":"200"}
{"level":"debug","ts":1615890053.484047,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"GET","url":"/api/v4/analytics/old","request_id":"kud7egk6dtgfzn6txz8dfhxjgw","status_code":"200"}
{"level":"debug","ts":1615890053.5056386,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"GET","url":"/api/v4/config","request_id":"zmikti8w9irttge6xwu1hszctc","status_code":"200"}
{"level":"debug","ts":1615890053.5106087,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"GET","url":"/api/v4/config/environment","request_id":"78osau1hmjd9tdf1b4keyw3pxa","status_code":"200"}
{"level":"debug","ts":1615890053.5352232,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"GET","url":"/api/v4/license/client","request_id":"9esumrtt6pbb7m9g95fndnzwoe","status_code":"200"}
{"level":"debug","ts":1615890053.8474827,"caller":"app/analytics.go:29","msg":"More than limit users are on the system, intensive queries skipped","limit":2500}
{"level":"debug","ts":1615890053.8656566,"caller":"web/handlers.go:103","msg":"Received HTTP request","method":"GET","url":"/api/v4/analytics/old","request_id":"qqnjxq91ppfupe49zz3xfixhba","status_code":"200"}
```

**Comment by streamer45** [comment]
Ran an additional load-test at high scale (25K concurrent users) which didn't show anything significant (all metrics well under their thresholds). 
cc @agnivade

**Comment by agnivade** [comment]
Wow, good stuff. Thanks a lot Claudio!

**Pull request merged**
